overlord/devicestate/devicestate.go

```diff
@@ -462,298 +418,385 @@ func modelSnapChannelFromDefaultOrPinnedTrack(new *asserts.Model, s *asserts.Mod
 // pass both the snap name and the model snap, as it is possible that
 // the model snap is nil for UC16 models
 type modelSnapsForRemodel struct {
-	currentSnap            string
-	currentModelSnap       *asserts.ModelSnap
-	new                    *asserts.Model
-	newSnap                string
-	newModelSnap           *asserts.ModelSnap
-	newRequiredRevision    snap.Revision
-	newModelValidationSets *snapasserts.ValidationSets
-}
+	new *asserts.Model
+
+	oldSnap      string
+	oldModelSnap *asserts.ModelSnap
 
-func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
-	return ms.newModelSnap != nil &&
-		(ms.newModelSnap.SnapType == "kernel" || ms.newModelSnap.SnapType == "gadget")
+	newSnap      string
+	newModelSnap *asserts.ModelSnap
 }
 
-type remodelVariant struct {
-	offline        bool
-	localSnaps     []*snap.SideInfo
-	localSnapPaths []string
+type remodeler struct {
```

Comment:
it might be worth in a follow to move this and its dedicated helpers to a remodeler.go file?